File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/client/playback/engine/specs/GivenAPlayingPlaybackEngine/AndAPreparationErrorOccurs/WhenObservingPlayback.java
package com.lasthopesoftware.bluewater.client.playback.engine.specs.GivenAPlayingPlaybackEngine.AndAPreparationErrorOccurs;

import com.lasthopesoftware.bluewater.client.browsing.items.media.files.ServiceFile;
import com.lasthopesoftware.bluewater.client.browsing.library.access.ILibraryStorage;
import com.lasthopesoftware.bluewater.client.browsing.library.access.ISpecificLibraryProvider;
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library;
import com.lasthopesoftware.bluewater.client.playback.engine.PlaybackEngine;
import com.lasthopesoftware.bluewater.client.playback.engine.bootstrap.PlaylistPlaybackBootstrapper;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.IPlayableFilePreparationSourceProvider;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparationException;
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlaybackQueueResourceManagement;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PreparedPlayableFile;
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CompletingFileQueueProvider;
import com.lasthopesoftware.bluewater.client.playback.file.specs.fakes.FakePreparedPlayableFile;
import com.lasthopesoftware.bluewater.client.playback.file.specs.fakes.ResolvablePlaybackHandler;
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.storage.NowPlaying;
import com.lasthopesoftware.bluewater.client.playback.view.nowplaying.storage.NowPlayingRepository;
import com.lasthopesoftware.bluewater.client.playback.volume.PlaylistVolumeManager;
import com.lasthopesoftware.bluewater.shared.promises.extensions.specs.FuturePromise;
import com.namehillsoftware.handoff.Messenger;
import com.namehillsoftware.handoff.promises.Promise;

import org.jetbrains.annotations.NotNull;
import org.junit.BeforeClass;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.concurrent.ExecutionException;

import static org.assertj.core.api.Assertions.assertThat;

public class WhenObservingPlayback {

	private static PreparationException error;
	private static NowPlaying nowPlaying;

	@BeforeClass
	public static void context() throws InterruptedException, ExecutionException {
		final DeferredErrorPlaybackPreparer deferredErrorPlaybackPreparer = new DeferredErrorPlaybackPreparer();

		final IPlayableFilePreparationSourceProvider fakePlaybackPreparerProvider = new IPlayableFilePreparationSourceProvider() {
			@Override
			public PlayableFilePreparationSource providePlayableFilePreparationSource() {
				return deferredErrorPlaybackPreparer;
			}

			@Override
			public int getMaxQueueSize() {
				return 1;
			}
		};

		final Library library = new Library();
		library.setId(1);

		final ISpecificLibraryProvider libraryProvider = () -> new Promise<>(library);
		final ILibraryStorage libraryStorage = new ILibraryStorage() {
			@NotNull
			@Override
			public Promise<Library> saveLibrary(@NotNull Library library) {
				return new Promise<>(library);
			}

			@NotNull
			@Override
			public Promise<?> removeLibrary(@NotNull Library library) {
				return Promise.empty();
			}
		};

		final NowPlayingRepository nowPlayingRepository = new NowPlayingRepository(libraryProvider, libraryStorage);
		final PlaybackEngine playbackEngine = new FuturePromise<>(PlaybackEngine.createEngine(
			new PreparedPlaybackQueueResourceManagement(
				fakePlaybackPreparerProvider,
				fakePlaybackPreparerProvider),
			Collections.singletonList(new CompletingFileQueueProvider()),
			nowPlayingRepository,
			new PlaylistPlaybackBootstrapper(new PlaylistVolumeManager(1.0f)))).get();

		new FuturePromise<>(playbackEngine
			.setOnPlaylistError(e -> {
				if (e instanceof PreparationException)
					error = (PreparationException)e;
			})
			.startPlaylist(
				Arrays.asList(
					new ServiceFile(1),
					new ServiceFile(2),
					new ServiceFile(3),
					new ServiceFile(4),
					new ServiceFile(5)),
				0,
				0))
			.get();

		deferredErrorPlaybackPreparer.resolve().resolve();
		deferredErrorPlaybackPreparer.reject();

		nowPlaying = new FuturePromise<>(nowPlayingRepository.getNowPlaying()).get();
	}

	@Test
	public void thenTheErrorIsBroadcast() {
		assertThat(error).isNotNull();
	}

	@Test
	public void thenTheFilePositionIsSaved() {
		assertThat(nowPlaying.playlistPosition).isEqualTo(1);
	}

	private static class DeferredErrorPlaybackPreparer implements PlayableFilePreparationSource {

		private Messenger<PreparedPlayableFile> messenger;

		ResolvablePlaybackHandler resolve() {
			final ResolvablePlaybackHandler playbackHandler = new ResolvablePlaybackHandler();
			if (messenger != null)
				messenger.sendResolution(new FakePreparedPlayableFile<>(playbackHandler));

			return playbackHandler;
		}

		void reject() {
			if (messenger != null)
				messenger.sendRejection(new Exception());
		}

		@Override
		public Promise<PreparedPlayableFile> promisePreparedPlaybackFile(ServiceFile serviceFile, long preparedAt) {
			return new Promise<>(messenger -> this.messenger = messenger);
		}
	}
}

File: projectBlueWater/src/test/java/com/lasthopesoftware/bluewater/client/playback/engine/GivenAPlayingPlaybackEngine/AndAPreparationErrorOccurs/WhenObservingPlayback.kt
package com.lasthopesoftware.bluewater.client.playback.engine.GivenAPlayingPlaybackEngine.AndAPreparationErrorOccurs

import com.lasthopesoftware.bluewater.client.browsing.files.ServiceFile
import com.lasthopesoftware.bluewater.client.browsing.library.access.FakeLibraryRepository
import com.lasthopesoftware.bluewater.client.browsing.library.repository.Library
import com.lasthopesoftware.bluewater.client.browsing.library.repository.LibraryId
import com.lasthopesoftware.bluewater.client.connection.selected.GivenANullConnection.AndTheSelectedLibraryChanges.FakeSelectedLibraryProvider
import com.lasthopesoftware.bluewater.client.playback.engine.PlaybackEngine
import com.lasthopesoftware.bluewater.client.playback.engine.bootstrap.ManagedPlaylistPlayer
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.IPlayableFilePreparationSourceProvider
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparationException
import com.lasthopesoftware.bluewater.client.playback.engine.preparation.PreparedPlaybackQueueResourceManagement
import com.lasthopesoftware.bluewater.client.playback.file.fakes.FakePreparedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.fakes.ResolvablePlaybackHandler
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PlayableFilePreparationSource
import com.lasthopesoftware.bluewater.client.playback.file.preparation.PreparedPlayableFile
import com.lasthopesoftware.bluewater.client.playback.file.preparation.queues.CompletingFileQueueProvider
import com.lasthopesoftware.bluewater.client.playback.nowplaying.storage.NowPlaying
import com.lasthopesoftware.bluewater.client.playback.nowplaying.storage.NowPlayingRepository
import com.lasthopesoftware.bluewater.client.playback.volume.PlaylistVolumeManager
import com.lasthopesoftware.bluewater.shared.promises.extensions.toExpiringFuture
import com.namehillsoftware.handoff.Messenger
import com.namehillsoftware.handoff.promises.Promise
import org.assertj.core.api.Assertions.assertThat
import org.joda.time.Duration
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test

private const val libraryId = 668

class WhenObservingPlayback {
	private val mut by lazy {
		val deferredErrorPlaybackPreparer = DeferredErrorPlaybackPreparer()
		val fakePlaybackPreparerProvider: IPlayableFilePreparationSourceProvider =
			object : IPlayableFilePreparationSourceProvider {
				override fun providePlayableFilePreparationSource(): PlayableFilePreparationSource {
					return deferredErrorPlaybackPreparer
				}

                override val maxQueueSize: Int
                    get() {
                        return 1
                    }
			}

		val library = Library(id = libraryId)
		val libraryProvider = FakeLibraryRepository(library)

		val nowPlayingRepository = NowPlayingRepository(
			FakeSelectedLibraryProvider(),
			libraryProvider,
		)

		val preparedPlaybackQueueResourceManagement = PreparedPlaybackQueueResourceManagement(
			fakePlaybackPreparerProvider,
			fakePlaybackPreparerProvider
		)
		val playbackBootstrapper = ManagedPlaylistPlayer(
			PlaylistVolumeManager(1.0f),
			preparedPlaybackQueueResourceManagement,
			nowPlayingRepository,
			listOf(CompletingFileQueueProvider()),
		)
		val playbackEngine = PlaybackEngine(
			preparedPlaybackQueueResourceManagement,
			listOf(CompletingFileQueueProvider()),
			nowPlayingRepository,
			playbackBootstrapper,
			playbackBootstrapper,
		)

		Triple(deferredErrorPlaybackPreparer, nowPlayingRepository, playbackEngine)
	}

	private var error: PreparationException? = null
	private var nowPlaying: NowPlaying? = null

	@BeforeAll
	fun act() {
		val (deferredErrorPlaybackPreparer, nowPlayingRepository, playbackEngine) = mut

		val promisedError = Promise {
			playbackEngine
				.setOnPlaylistError { e ->
					if (e is PreparationException) it.sendResolution(e)
				}
		}

		playbackEngine
			.startPlaylist(
				LibraryId(libraryId),
				listOf(
					ServiceFile("1"),
					ServiceFile("2"),
					ServiceFile("3"),
					ServiceFile("4"),
					ServiceFile("5")
				),
				0
			)
			.toExpiringFuture()
			.get()

		deferredErrorPlaybackPreparer.resolve().resolve()
		deferredErrorPlaybackPreparer.reject()

		error = promisedError.toExpiringFuture().get()
		nowPlaying = nowPlayingRepository.promiseNowPlaying(LibraryId(libraryId)).toExpiringFuture().get()
	}

	@Test
	fun `then the error is broadcast`() {
		assertThat(error).isNotNull
	}

	@Test
	fun `then the file position is saved`() {
		assertThat(nowPlaying!!.playlistPosition).isEqualTo(1)
	}

	private class DeferredErrorPlaybackPreparer : PlayableFilePreparationSource {
		private var messenger: Messenger<PreparedPlayableFile?>? = null
		fun resolve(): ResolvablePlaybackHandler {
			val playbackHandler = ResolvablePlaybackHandler()
			messenger?.sendResolution(
				FakePreparedPlayableFile(
					playbackHandler
				)
			)
			return playbackHandler
		}
		fun reject() {
			messenger?.sendRejection(Exception())
		}

		override fun promisePreparedPlaybackFile(
			libraryId: LibraryId,
			serviceFile: ServiceFile,
			preparedAt: Duration
		): Promise<PreparedPlayableFile?> {
			return Promise { messenger ->
				this.messenger = messenger
			}
		}
	}
}
